toolkit/tools/scheduler/schedulerutils/preparerequest.go

```diff
@@ -23,51 +23,149 @@ func ConvertNodesToRequests(pkgGraph *pkggraph.PkgGraph, graphMutex *sync.RWMute
-			IsDelta:        hasADeltaNode,
-		}
+		defaultNode := buildNodes[0]
+		req := buildRequest(pkgGraph, buildState, packagesToRebuild, defaultNode, buildNodes, isCacheAllowed, hasADeltaNode)
 
-		req.CanUseCache = isCacheAllowed && canUseCacheForNode(pkgGraph, req.Node, packagesToRebuild, buildState)
+		if req.UseCache {
+			expectedFiles, missingFiles := pkggraph.FindRPMFiles(defaultNode.SrpmPath, pkgGraph, nil)
+			if len(missingFiles) > 0 && len(missingFiles) < len(expectedFiles) {
+				logger.Log.Infof("SRPM '%s' will be rebuilt due to partially missing components: %v", defaultNode.SRPMFileName(), missingFiles)
+			}
+
+			req.ExpectedFiles = expectedFiles
+			req.UseCache = len(missingFiles) == 0
+		}
 
 		requests = append(requests, req)
+
+		partnerTestNodeRequest := partnerTestNodesToRequest(pkgGraph, buildState, packagesToRebuild, buildNodes, req.UseCache)
+		if partnerTestNodeRequest != nil {
+			requests = append(requests, partnerTestNodeRequest)
+		}
+	}
+
+	return
+}
+
+func buildNodeToTestNode(pkgGraph *pkggraph.PkgGraph, buildNode *pkggraph.PkgNode) (testNode *pkggraph.PkgNode) {
+	dependents := pkgGraph.To(buildNode.ID())
+	for dependents.Next() {
+		dependent := dependents.Node().(*pkggraph.PkgNode)
+
+		if dependent.Type == pkggraph.TypeTest {
+			testNode = dependent
+			break
+		}
+	}
+
+	return
+}
+
+func buildRequest(pkgGraph *pkggraph.PkgGraph, buildState *GraphBuildState, packagesToRebuild []*pkgjson.PackageVer, builtNode *pkggraph.PkgNode, ancillaryNodes []*pkggraph.PkgNode, isCacheAllowed, isDelta bool) (request *BuildRequest) {
+	request = &BuildRequest{
+		Node:           builtNode,
+		PkgGraph:       pkgGraph,
+		AncillaryNodes: ancillaryNodes,
+		IsDelta:        isDelta,
+	}
+
+	request.UseCache = isCacheAllowed && canUseCacheForNode(pkgGraph, request.Node, packagesToRebuild, buildState)
+	return
+}
+
+func partnerTestNodesToRequest(pkgGraph *pkggraph.PkgGraph, buildState *GraphBuildState, packagesToRebuild []*pkgjson.PackageVer, buildNodes []*pkggraph.PkgNode, buildUsesCache bool) (request *BuildRequest) {
+	const isDelta = false
+
+	defaultBuildNode := buildNodes[0]
+	testNode := buildNodeToTestNode(pkgGraph, defaultBuildNode)
+	if testNode == nil {
+		return
+	}
+
+	ancillaryTestNodes := []*pkggraph.PkgNode{}
+	for _, buildNode := range buildNodes {
+		testNode = buildNodeToTestNode(pkgGraph, buildNode)
+
+		// Removing edges even if tests are blocked by other dependencies,
+		// so that they can get unblocked once these other dependencies are available.
+		pkgGraph.RemoveEdge(testNode.ID(), buildNode.ID())
+
+		ancillaryTestNodes = append(ancillaryTestNodes, testNode)
+	}
+
+	if !isNodeUnblocked(pkgGraph, buildState, testNode) {
+		return
+	}
+
+	return &BuildRequest{
+		Node:           testNode,
+		PkgGraph:       pkgGraph,
+		AncillaryNodes: ancillaryTestNodes,
+		IsDelta:        isDelta,
+		UseCache:       buildUsesCache && canUseCacheForNode(pkgGraph, testNode, packagesToRebuild, buildState),
```

Comment:
Do we need to grab the value of `canUseCacheForNode()` prior to deleting the edges? That function checks rebuliding dependencies based on those edges, and I think the goal here is to ensure we re-run the tests in the even the build node is triggered to rebuild?

Reply:
This is tricky, but we need to call `canUseCacheForNode()` after removing that edge. Since the partner build node is not built yet, then this edge would make `canUseCacheForNode()` always return `false`. At this point the partner build node is being sent to the build worker, so it won't get analyzed again in the future.

In other word the process for the test nodes is:
- Hold it's analysis with the virtual B -> T edge.
- Once the build node is unblocked, analyze its partner test node at the same time. That's this function here. We remove the virtual edge and the test node either gets immediately queued or is blocked on some extra dependencies - that's the part decided by `canUseCacheForNode()`.
- If the test node ends up being blocked, it gets re-analyzed later once these other dependencies are done. These unblocked nodes end up inside `testNodes` from `ConvertNodesToRequests()` and are queued for building in the separate `testNodesToRequests()` function. At this point the partner build nodes for these test nodes have either already finished building or are being built, thus the check for active and cached SRPMs inside that function.

Reply:
Ok, I think I follow. Can we add something similar to the code as a comment for future me, who will also undoubtedly become confused?

Reply:
Very reasonable request, I should have added it anyway. Thanks!